tt_eager/tt_dnn/op_library/sdpa/multi_core/sdpa_decode_op_multi_core.cpp

```diff
@@ -0,0 +1,628 @@
+namespace primary {
+
+// implementation of softmax with optional scale/mask (see the header for input_tensor more detailed description)
+operation::ProgramWithCallbacks sdpa_decode_multi_core(
+    const Tensor &input_tensor_q,
+    const Tensor &input_tensor_k,
+    const Tensor &input_tensor_v,
+    const Tensor &output_tensor,
+    const std::optional<const Tensor> attn_mask,
+    std::optional<float> scale,
+    std::size_t k_chunk_size,
+    DeviceComputeKernelConfig compute_kernel_config,
+    transformers::SDPAProgramConfig program_config,
+    std::optional<const uint32_t> valid_seq_len
+) {
+
+    /*
+    Q: 1 x B x PNH x DH
+    K: 1 x B x S x DH
+    V: 1 x B x S x DH
+    attn_mask: 1 x B x PNH x S
+    */
+
+    const auto q_shape = input_tensor_q.get_legacy_shape();
+    const auto k_shape = input_tensor_k.get_legacy_shape();
+    // Use k_shape for S and DH since Q might be different for decode
+    uint32_t B = q_shape[1], PNH = q_shape[2], S = k_shape[2], DH = k_shape[3];
+    uint32_t PSt = valid_seq_len.value()/TILE_HEIGHT;
+    uint32_t St = S/TILE_HEIGHT;
+    uint32_t DHt = DH/TILE_WIDTH;
+    uint32_t PNHt = PNH/TILE_HEIGHT;
+    uint32_t Sk_chunk_t = k_chunk_size / TILE_HEIGHT;
+    uint32_t num_chunks = valid_seq_len.value() / k_chunk_size;
+    bool is_q_sharded = input_tensor_q.is_sharded();
+    bool is_output_sharded = output_tensor.is_sharded();
+
+    // log_debug all of the above
+    log_debug("B: {}", B);
+    log_debug("S: {}", S);
+    log_debug("DH: {}", DH);
+    log_debug("St: {}", St);
+    log_debug("DHt: {}", DHt);
+    log_debug("PNHt: {}", PNHt);
+    log_debug("Sk_chunk_t: {}", Sk_chunk_t);
+    log_debug("k_chunk_size: {}", k_chunk_size);
+    log_debug("num_chunks: {}", num_chunks);
+
+    Program program = CreateProgram();
+
+    Device *device = input_tensor_q.device();
+
+    MathFidelity math_fidelity;
+    bool math_approx_mode;
+    bool fp32_dest_acc_en;
+
+
+    std::visit([&](auto&& compute_kernel_config) {
+        using T = std::decay_t<decltype(compute_kernel_config)>;
+        if constexpr (std::is_same_v<T, GrayskullComputeKernelConfig>) {
+            TT_FATAL(device->arch() == ARCH::GRAYSKULL, "kernel config is not for graykull");
+            math_fidelity = compute_kernel_config.math_fidelity;
+            math_approx_mode = compute_kernel_config.math_approx_mode;
+            fp32_dest_acc_en = false;
+        } else if constexpr (std::is_same_v<T, WormholeComputeKernelConfig>) {
+            TT_FATAL(device->arch() == ARCH::WORMHOLE_B0, "kernel config is not for wormhole_b0");
+            math_fidelity = compute_kernel_config.math_fidelity;
+            math_approx_mode = compute_kernel_config.math_approx_mode;
+            fp32_dest_acc_en = compute_kernel_config.fp32_dest_acc_en;
+        } else {
+            TT_FATAL("arch not supported");
+        }
+
+    }, compute_kernel_config);
+
+    // TT_FATAL(!fp32_dest_acc_en, "fp32_dest_acc_en not supported yet");
```

Comment:
Do we check somewhere to disallow fp32 accumulate?

Reply:
Yes. I check it in the validate function. I will remove it here